KAFKA-10199: Add RESUME in state updater

guozhangwang opened the pull request:
This should be reviewed after https://github.com/apache/kafka/pull/12386.

1) Need to check `enforceRestoreActive` / `transitToUpdateStandby` when resuming a paused task.
2) Do not expose another `getResumedTasks` since I think its caller only need the `getPausedTasks`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

guozhangwang committed:
commit during restoration

guozhangwang committed:
more unit tests

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K10199-commit-in-state-updater

guozhangwang committed:
first commit

guozhangwang committed:
rebased from trunk

guozhangwang committed:
github comments from 12279

guozhangwang committed:
remove unused function

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K10199-remove-main-consumer

guozhangwang committed:
add requireStable to admin client request option

guozhangwang committed:
remove unused imports

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K10199-remove-main-consumer

guozhangwang committed:
add paused tasks

guozhangwang committed:
add unit tests

guozhangwang committed:
add more unit tests

guozhangwang committed:
add unit tests

guozhangwang committed:
rebase from trunk

guozhangwang committed:
remove func

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K10199-pause-tasks-in-updater

guozhangwang committed:
checkstyle

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskAndAction.java:
```diff
@@ -47,6 +49,16 @@ public static TaskAndAction createRemoveTask(final TaskId taskId) {
+        return new TaskAndAction(null, taskId, Action.PAUSE);
+    }
+
+    public static TaskAndAction createResumeTask(final TaskId taskId) {
+        Objects.requireNonNull(taskId, "Task ID of task to pause is null!");
+        return new TaskAndAction(null, taskId, Action.RESUME);
+    }
```
Could you please add a unit test for this?

guozhangwang replied:
ack.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskAndAction.java:
```diff
@@ -55,7 +67,7 @@ public Task getTask() {
 
     public TaskId getTaskId() {
-        if (action != Action.REMOVE) {
+        if (action != Action.REMOVE && action != Action.PAUSE && action != Action.RESUME) {
```
Could you please add a unit test for this?

guozhangwang replied:
Ack.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdaterTest.java:
```diff
@@ -428,6 +456,113 @@ private void shouldRemoveStatefulTask(final Task task) throws Exception {
+    }
+
+    @Test
+    public void shouldRemovePausedTask() throws Exception {
+        final StreamTask task = createActiveStatefulTaskInStateRestoring(TASK_0_0, Collections.singletonList(TOPIC_PARTITION_A_0));
+        stateUpdater.start();
+        stateUpdater.add(task);
+
+        stateUpdater.pause(task.id());
+
+        verifyPausedTasks(task);
+        verifyCheckpointTasks(true, task);
+        verifyRestoredActiveTasks();
+        verifyRemovedTasks();
+        verifyUpdatingTasks();
+
+        stateUpdater.remove(task.id());
+
+        verifyPausedTasks();
+        verifyRemovedTasks(task);
+        verifyRestoredActiveTasks();
+        verifyUpdatingTasks();
+
+        stateUpdater.resume(task.id());
+
+        verifyPausedTasks();
+        verifyRemovedTasks(task);
+        verifyRestoredActiveTasks();
+        verifyUpdatingTasks();
+    }
```
Do we really need the part from line 558 until the end? 
You could subdivide this test into `shouldRemovePausedTask()` and `shouldNotResumeRemovedTask()`. The former is basically this test until line 556 and the latter is to test whether a removed task can be resumed. Then we should also consider to write tests to verify if failed and restored tasks cannot be resumed.

guozhangwang replied:
Makes sense!

guozhangwang replied:
I added the other tests but when trying to cover `shouldNotResumeActiveStatefulTaskInFailedTasks` and `shouldNotResumeStandbyTaskInFailedTasks` I realize they should not happen: this is because we should never fail an non-updating (i.e. paused) task, and in fact we have a check in our code when `handleStreamsExceptionWithTask`, so I removed these tests.

cadonna replied:
I was rather thinking about verifying that removed tasks, failed tasks, and restored tasks cannot be resumed. That means without pausing them beforehand. You just ensure that a task failed and then you try to resume it. The same applies to restored and removed tasks.

guozhangwang replied:
Sounds good, ack.

cadonna reviewed (commented):
@guozhangwang Thank you for the PR!

Here my feedback.

guozhangwang committed:
github comments

guozhangwang commented:
Thanks @cadonna ! I incorporated your comments and also added more tests. Since I've not merged the previous one the diff should be clearer to review now.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdaterTest.java:
```diff
@@ -723,6 +723,106 @@ private void shouldNotPauseTaskInRemovedTasks(final Task task) throws Exception
+    }
+
+    @Test
+    public void shouldIgnoreResumingNotPausedTasks() throws Exception {
```
I would rather call this `shouldNotResumeNotExistingTasks()`.

guozhangwang replied:
ack.

cadonna reviewed (approved):
Thanks for the PR, @guozhangwang !

I had two comments about tests.

For the rest, it LGTM!

guozhangwang committed:
github comments

guozhangwang commented:
Thanks @cadonna , I've incorporated your comments and merged to trunk.
